File: September2022/Java/DesignCircularQueue.java

// @saorav21994

class MyCircularQueue {
    
    // Queue -> add in rear and delete in front -> FIFO
    
    LinkedList<Integer> queue;
    int size;
    int cur = 0;
    public MyCircularQueue(int k) {
        size = k;
        queue = new LinkedList<>();
    }
    
    public boolean enQueue(int value) {
        if (cur == size) 
            return false;   // queue full
        cur += 1;
        queue.addLast(value);
        return true;
    }
    
    public boolean deQueue() {
        if (cur == 0)
            return false;   //queue empty
        cur -= 1;
        queue.removeFirst();
        return true;
    }
    
    public int Front() {
        if (cur == 0)
            return -1;
        return queue.getFirst();
    }
    
    public int Rear() {
        if (cur == 0)
            return -1;
        return queue.getLast();    
    }
    
    public boolean isEmpty() {
        return (cur == 0);
    }
    
    public boolean isFull() {
        return (cur == size);
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue obj = new MyCircularQueue(k);
 * boolean param_1 = obj.enQueue(value);
 * boolean param_2 = obj.deQueue();
 * int param_3 = obj.Front();
 * int param_4 = obj.Rear();
 * boolean param_5 = obj.isEmpty();
 * boolean param_6 = obj.isFull();
 */

File: September2022/C++/DesignCircularQueue.cpp
// https://leetcode.com/problems/design-circular-queue/
class MyCircularQueue {
private:
    vector<int> queue;
    int front, rear, length;
    
public:
    // SC: O(k)
    MyCircularQueue(int k) {
        vector<int> temp(k, 0);
        queue = temp;
        front = 0;
        rear = -1;
        length = 0;
    }
    
    // TC: O(1)
    bool enQueue(int value) {
        if (isFull()) {
            return false;
        }
        
        rear += 1;
        rear %= queue.size();   // alternate -> rear = (rear + 1) % queue.size();
        queue[rear] = value;
        length++;
        return true;
    }
    
    // TC: O(1)
    bool deQueue() {
        if (isEmpty()) {
            return false;
        }
        queue[front] = -1;
        front += 1;
        front %= queue.size();
        length--;
        return true;
    }
    
    // TC: O(1)
    int Front() {
        return !isEmpty() ? queue[front] : -1;
    }
    
    // TC: O(1)
    int Rear() {
        return !isEmpty() ? queue[rear] : -1;
    }
    
    // TC: O(1)
    bool isEmpty() {
        return length == 0;
    }
    
    // TC: O(1)
    bool isFull() {
        return length == queue.size();
    }
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */
